src/validation/caseInsensitiveDocument.ts

```diff
@@ -3,10 +3,33 @@ export type Recordable<T> = {
 };
 
 export class CaseInsensitiveDocument<T> {
-  public documentRecord: Record<string, T> = {};
+  private _originalDocument: Record<string, T> = {};
+  private _remainingRecord: Record<string, T> = {};
+
   public constructor(doc: Record<string, T>) {
+    this._originalDocument = this.shallowCopy(doc);
```

Comment:
This class needs to always keep a copy of the original record so it can be used for error messages

Reply:
nitpik, I think for object copy, this two methods are clearer/preferable:
- Shallow : `const copy = Object.create(original)`
- Deep: `const copy = structuredClone(original)`

Note, `structuredClone` is only available since Node 17.

Reply:
Not exactly the same thing. Correct me if I'm wrong, but `Object.create` only creates references as opposed to `Object.assign` which copies all enumerable properties at the first level from the source object.